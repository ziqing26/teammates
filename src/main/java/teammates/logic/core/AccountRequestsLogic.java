package teammates.logic.core;

import teammates.common.datatransfer.attributes.AccountRequestAttributes;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.storage.api.AccountRequestsDb;

/**
 * Handles the logic related to account requests.
 */
public final class AccountRequestsLogic {

    private static final AccountRequestsLogic instance = new AccountRequestsLogic();

    private final AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

    private AccountRequestsLogic() {
        // prevent initialization
    }

    public static AccountRequestsLogic inst() {
        return instance;
    }

    void initLogicDependencies() {
        // No dependency to other logic class
    }

    /**
     * Updates an account request.
     *
     * @return the updated account request
     * @throws InvalidParametersException if the account request is not valid
     * @throws EntityDoesNotExistException if the account request to create does not exist
     */
    public AccountRequestAttributes updateAccountRequest(AccountRequestAttributes.UpdateOptions updateOptions)
            throws InvalidParametersException, EntityDoesNotExistException {
        return accountRequestsDb.updateAccountRequest(updateOptions);
    }

    /**
     * Creates an account request.
     *
     * @return the created account request
     * @throws InvalidParametersException if the account request is not valid
     * @throws EntityAlreadyExistsException if the account request to create already exists
     */
    public AccountRequestAttributes createAccountRequest(AccountRequestAttributes accountRequest)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return accountRequestsDb.createEntity(accountRequest);
    }

    /**
     * Deletes the account request associated with the email address and institute.
     *
     * <p>Fails silently if the account request doesn't exist.</p>
     */
    public void deleteAccountRequest(String email, String institute) {
        accountRequestsDb.deleteAccountRequest(email, institute);
    }

    /**
     * Gets an account request by email address and institute.
     *
     * @return the account request
     */
    public AccountRequestAttributes getAccountRequest(String email, String institute) {
        return accountRequestsDb.getAccountRequest(email, institute);
    }

    /**
     * Gets an account request by unique constraint {@code registrationKey}.
     *
     * @return the account request
     */
    public AccountRequestAttributes getAccountRequestForRegistrationKey(String registrationKey) {
        return accountRequestsDb.getAccountRequestForRegistrationKey(registrationKey);
    }

}
