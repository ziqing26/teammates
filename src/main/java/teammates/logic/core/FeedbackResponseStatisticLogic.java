package teammates.logic.core;

import java.time.Duration;
import java.time.Instant;
import java.util.List;

import teammates.common.datatransfer.attributes.FeedbackResponseStatisticAttributes;
import teammates.common.util.Const;
import teammates.storage.api.FeedbackResponseStatisticsDb;

/**
 * Handles operations related to feedback response statistics.
 */
public class FeedbackResponseStatisticLogic {

	private static final FeedbackResponseStatisticLogic instance = new FeedbackResponseStatisticLogic();

	private final FeedbackResponseStatisticsDb feedbackResponseStatisticDb = FeedbackResponseStatisticsDb.inst();

	private FeedbackResponseStatisticLogic() {
		// prevent initialization
	}
	
	public static FeedbackResponseStatisticLogic inst() {
		return instance;
	}

	/**
     * Gets all feedback response statistics in time period.
     */
    public List<FeedbackResponseStatisticAttributes> getFeedbackResponseStatistics(Instant startTime, Instant endTime) {

		if (Duration.between(startTime, endTime).compareTo(Const.FEEDBACK_STATISTIC_MINUTE_THRESHHOLD) <= 0) {
			return feedbackResponseStatisticDb.getFeedbackResponseStatisticsInInterval(startTime, endTime, Const.MINUTE_IN_SECONDS);
		}

		return feedbackResponseStatisticDb.getFeedbackResponseStatisticsInInterval(startTime, endTime, Const.HOUR_IN_SECONDS);
    }
}
