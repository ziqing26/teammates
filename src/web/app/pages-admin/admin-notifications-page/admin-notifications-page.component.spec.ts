import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import moment from 'moment-timezone';
import { of, throwError } from 'rxjs';
import SpyInstance = jest.SpyInstance;
import { NotificationService } from '../../../services/notification.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import { createMockNgbModalRef } from '../../../test-helpers/mock-ngb-modal-ref';
import { Notification, NotificationStyle, NotificationTargetUser } from '../../../types/api-output';
import { SortBy } from '../../../types/sort-properties';
import { SimpleModalType } from '../../components/simple-modal/simple-modal-type';
import { AdminNotificationsPageComponent } from './admin-notifications-page.component';
import { AdminNotificationsPageModule } from './admin-notifications-page.module';
import { NotificationEditFormModel } from './notification-edit-form/notification-edit-form-model';
import { NotificationsTableRowModel } from './notifications-table/notifications-table-model';

const testNotificationEditModel: NotificationEditFormModel = {
  notificationId: 'notification1',
  shown: false,

  startTime: { hour: 0, minute: 0 },
  startDate: { year: 0, month: 0, day: 0 },
  endTime: { hour: 0, minute: 0 },
  endDate: { year: 0, month: 0, day: 0 },

  style: NotificationStyle.SUCCESS,
  targetUser: NotificationTargetUser.INSTRUCTOR,

  title: 'valid title',
  message: 'valid message',

  isSaving: false,
  isDeleting: false,
};

const testNotificationOne: Notification = {
  notificationId: 'notification1',
  startTimestamp: moment('2017-09-15 09:30:00').valueOf(),
  endTimestamp: moment('2050-09-15 09:30:00').valueOf(),
  createdAt: moment('2017-09-15 09:30:00').valueOf(),
  style: NotificationStyle.SUCCESS,
  targetUser: NotificationTargetUser.INSTRUCTOR,
  title: 'valid title 1',
  message: 'valid message 1',
  shown: false,
};

const testNotificationTwo: Notification = {
  notificationId: 'notification2',
  startTimestamp: moment('2018-12-15 09:30:00').valueOf(),
  endTimestamp: moment('2050-09-15 09:30:00').valueOf(),
  createdAt: moment('2018-11-15 09:30:00').valueOf(),
  style: NotificationStyle.DANGER,
  targetUser: NotificationTargetUser.GENERAL,
  title: 'valid title 2',
  message: 'valid message 2',
  shown: false,
};

const notificationTableRowModel1: NotificationsTableRowModel = {
  isHighlighted: true,
  notification: testNotificationOne,
};

const notificationTableRowModel2: NotificationsTableRowModel = {
  isHighlighted: false,
  notification: testNotificationTwo,
};

describe('AdminNotificationsPageComponent', () => {
  let component: AdminNotificationsPageComponent;
  let fixture: ComponentFixture<AdminNotificationsPageComponent>;
  let notificationService: NotificationService;
  let statusMessageService: StatusMessageService;
  let simpleModalService: SimpleModalService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        AdminNotificationsPageModule,
        HttpClientTestingModule,
        BrowserAnimationsModule,
      ],
      providers: [
        TimezoneService,
      ],
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(AdminNotificationsPageComponent);
    notificationService = TestBed.inject(NotificationService);
    statusMessageService = TestBed.inject(StatusMessageService);
    simpleModalService = TestBed.inject(SimpleModalService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when notifications are loading', () => {
    component.isNotificationLoading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when notifications failed to load', () => {
    component.hasNotificationLoadingFailed = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when notification edit form expanded for adding notification', () => {
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of());
    component.initNotificationEditFormModel();
    component.loadNotifications();
    fixture.detectChanges();

    const btn: any = fixture.debugElement.nativeElement.querySelector('#btn-add-notification');
    expect(btn.disabled).toBeFalsy();
    expect(btn.textContent).toEqual(' Add New Notification');
    btn.click();
    fixture.detectChanges();

    expect(component.isNotificationLoading).toBeFalsy();
    expect(component.isNotificationEditFormExpanded).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should disable edit button when notification edit form expanded for existent notification', () => {
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of({
      notifications: [testNotificationOne],
    }));
    component.initNotificationEditFormModel();
    component.loadNotifications();
    component.loadNotificationEditForm(component.notificationsTableRowModels[0].notification);
    fixture.detectChanges();

    const btn: any = fixture.debugElement.nativeElement.querySelector('#btn-add-notification');
    expect(btn.disabled).toBeTruthy();
    expect(btn.textContent).toEqual(' Edit Existing Notification');
  });

  it('should load correct notification for a given API output', () => {
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of({
      notifications: [testNotificationOne],
    }));

    component.loadNotifications();

    expect(component.notificationsTableRowModels.length).toEqual(1);
    expect(component.notificationsTableRowModels[0].notification.notificationId)
      .toEqual(testNotificationOne.notificationId);
    expect(component.notificationsTableRowModels[0].notification.shown).toBeFalsy();
    expect(component.notificationsTableRowModels[0].notification.targetUser).toEqual(testNotificationOne.targetUser);
    expect(component.notificationsTableRowModels[0].notification.style).toEqual(testNotificationOne.style);
    expect(component.notificationsTableRowModels[0].notification.title).toEqual(testNotificationOne.title);
    expect(component.notificationsTableRowModels[0].notification.message).toEqual(testNotificationOne.message);
    expect(component.notificationsTableRowModels[0].notification.startTimestamp)
      .toEqual(testNotificationOne.startTimestamp);
    expect(component.notificationsTableRowModels[0].notification.endTimestamp)
      .toEqual(testNotificationOne.endTimestamp);
  });

  it('should add notification for all fields filled in', () => {
    jest.spyOn(notificationService, 'createNotification').mockReturnValue(of(testNotificationOne));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
    .mockImplementation((args: string) => {
      expect(args).toEqual('Successfully created');
    });

    component.addNewNotificationHandler();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.isNotificationEditFormExpanded).toBeFalsy();
    expect(component.notificationsTableRowModels.length).toEqual(1);
    expect(component.notificationsTableRowModels[0].notification.notificationId)
      .toEqual(testNotificationOne.notificationId);
    expect(component.notificationsTableRowModels[0].notification.shown).toBeFalsy();
    expect(component.notificationsTableRowModels[0].notification.style).toEqual(testNotificationOne.style);
    expect(component.notificationsTableRowModels[0].notification.targetUser).toEqual(testNotificationOne.targetUser);
    expect(component.notificationsTableRowModels[0].notification.title).toEqual(testNotificationOne.title);
    expect(component.notificationsTableRowModels[0].notification.message).toEqual(testNotificationOne.message);
    expect(component.notificationsTableRowModels[0].notification.startTimestamp)
      .toEqual(testNotificationOne.startTimestamp);
    expect(component.notificationsTableRowModels[0].notification.endTimestamp)
      .toEqual(testNotificationOne.endTimestamp);
  });

  it('should display error message when failed to create notification', () => {
    component.notificationEditFormModel = testNotificationEditModel;
    jest.spyOn(notificationService, 'createNotification').mockReturnValue(throwError({
      error: {
        message: 'This is the error message.',
      },
    }));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast')
      .mockImplementation((args: string) => {
        expect(args).toEqual('This is the error message.');
      });

    component.addNewNotificationHandler();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.notificationsTableRowModels.length).toEqual(0);
  });

  it('should display error message when notification failed to load', () => {
    component.hasNotificationLoadingFailed = false;
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(throwError({
      error: {
        message: 'This is the error message.',
      },
    }));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast')
      .mockImplementation((args: string) => {
        expect(args).toEqual('This is the error message.');
      });

    component.loadNotifications();

    expect(spy).toBeCalled();
    expect(component.hasNotificationLoadingFailed).toBeTruthy();
  });

  it('should display warning when attempts to edit another notification when form is open', async () => {
    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of({
      notifications: [testNotificationOne, testNotificationTwo],
    }));
    component.initNotificationEditFormModel();
    component.loadNotifications();
    component.loadNotificationEditForm(component.notificationsTableRowModels[0].notification);
    fixture.detectChanges();
    expect(component.isNotificationEditFormExpanded).toBeTruthy();

    const expectedMessage: string = component.notificationsTableRowModels[0].notification.message;

    component.loadNotificationEditFormHandler(component.notificationsTableRowModels[0].notification);
    component.notificationEditFormModel.message = 'new message';
    expect(component.notificationEditFormModel.message).toEqual('new message');

    // attempts to load edit form for another notification
    component.loadNotificationEditFormHandler(component.notificationsTableRowModels[1].notification);
    await promise;
    // the first loaded notification message should not be updated to new message
    expect(component.notificationsTableRowModels[0].notification.message).toEqual(expectedMessage);
    // notification edit form model should be updated to the newly loaded notification
    expect(component.notificationEditFormModel.message)
        .toEqual(component.notificationsTableRowModels[1].notification.message);

    expect(modalSpy).toHaveBeenCalledTimes(1);
    expect(modalSpy).toHaveBeenLastCalledWith('Discard unsaved edit?',
        SimpleModalType.WARNING,
        'Warning: If you choose to edit another notification, any unsaved changes will be lost.');
  });

  it('should delete notfication', () => {
    component.notificationsTableRowModels = [notificationTableRowModel1];
    expect(component.notificationsTableRowModels.length).toEqual(1);
    jest.spyOn(notificationService, 'deleteNotification').mockReturnValue(of({
      message: 'Successfully deleted',
    }));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
    .mockImplementation((args: string) => {
      expect(args).toEqual('Successfully deleted');
    });

    component.deleteNotificationHandler(testNotificationEditModel.notificationId);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.notificationsTableRowModels.length).toEqual(0);
  });

  it('should sort notification by start time', () => {
    component.notificationsTableRowModels = [notificationTableRowModel1, notificationTableRowModel2];
    component.notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_CREATE_TIME;
    component.sortNotificationsTableRowModelsHandler(SortBy.NOTIFICATION_START_TIME);
    expect(component.notificationsTableRowModelsSortBy).toEqual(SortBy.NOTIFICATION_START_TIME);

    const expected : NotificationsTableRowModel[] = [notificationTableRowModel2, notificationTableRowModel1]
        .sort(component.getNotificationsTableRowModelsComparator());
    expect(expected[0]).toBe(component.notificationsTableRowModels[0]);
    expect(expected[1]).toBe(component.notificationsTableRowModels[1]);
  });
});
