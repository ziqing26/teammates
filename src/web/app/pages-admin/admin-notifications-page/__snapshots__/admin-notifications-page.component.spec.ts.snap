// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminNotificationsPageComponent should snap when notification edit form expanded for adding notification 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  currentNotificationEditFormMode="0"
  guessTimezone={[Function String]}
  hasNotificationLoadingFailed="false"
  isNotificationEditFormExpanded={[Function Boolean]}
  isNotificationLoading="false"
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in Africa/Abidjan time.
    </p>
  </div><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <span>
        <i
          class="fas fa-plus"
        />
         Add New Notification
      </span>
    </button>
  </div><div
    class="margin-top-30px ng-trigger ng-trigger-collapseAnim ng-animating"
    style=""
  >
    <tm-notification-edit-form
      style=""
    >
      <div
        class="card card-plain"
      >
        <div
          class="close-header"
        >
          <button
            aria-label="Close"
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          class="card-body"
        >
          <div
            class="card border-primary margin-top-20px"
          >
            <div
              class="card-body"
            >
              <div
                class="row text-center"
              >
                <div
                  class="col-md-2 text-md-right font-weight-bold"
                >
                  <span
                    class="ngb-tooltip-class"
                    ngbtooltip="The indended recipients of this notification."
                  >
                    Target user group
                  </span>
                </div>
                <div
                  class="col-md-auto text-md-left"
                >
                  <select
                    class="form-control ng-untouched ng-pristine ng-valid"
                    id="notification-target-user"
                  >
                    <option
                      value="0: STUDENT"
                    >
                      Students
                    </option>
                    <option
                      value="1: INSTRUCTOR"
                    >
                      Instructors
                    </option>
                    <option
                      value="2: GENERAL"
                    >
                      General (for both students and instructors)
                    </option>
                  </select>
                </div>
              </div>
              <br />
              <div
                class="row text-center"
              >
                <div
                  class="col-md-2 text-md-right font-weight-bold"
                >
                  <span
                    class="ngb-tooltip-class"
                    ngbtooltip="Different styles correspond to different background colours at display."
                  >
                    Notification style
                  </span>
                </div>
                <div
                  class="col-md-auto text-md-left"
                >
                  <select
                    class="form-control ng-untouched ng-pristine ng-valid"
                    id="notification-style"
                  >
                    <option
                      value="0: DANGER"
                    >
                      Danger (red)
                    </option>
                    <option
                      value="1: DARK"
                    >
                      Dark
                    </option>
                    <option
                      value="2: INFO"
                    >
                      Info (cyan)
                    </option>
                    <option
                      value="3: LIGHT"
                    >
                      Light
                    </option>
                    <option
                      value="4: PRIMARY"
                    >
                      Primary (blue)
                    </option>
                    <option
                      value="5: SECONDARY"
                    >
                      Secondary (grey)
                    </option>
                    <option
                      value="6: SUCCESS"
                    >
                      Success (green)
                    </option>
                    <option
                      value="7: WARNING"
                    >
                      Warning (yellow)
                    </option>
                  </select>
                </div>
                <div
                  class="col-md-auto text-md-middle d-flex alert alert-success"
                >
                  <span
                    class="align-self-center"
                  >
                    Style Demonstration
                  </span>
                </div>
              </div>
              <br />
              <div
                class="row text-center"
              >
                <div
                  class="col-md-2 text-md-right font-weight-bold"
                >
                  <span
                    class="ngb-tooltip-class"
                    ngbtooltip="Enter the name of the notification e.g. Version Changelog 8.9.1."
                  >
                    Title
                  </span>
                </div>
                <div
                  class="col-md-10 text-md-left"
                >
                  <input
                    class="form-control ng-untouched ng-pristine ng-valid"
                    id="notification-title"
                    maxlength="80"
                    placeholder="e.g. Version Changelog 8.9.1"
                    type="text"
                  />
                  <div>
                     80 characters left 
                  </div>
                </div>
              </div>
              <br />
              <div
                class="row text-center"
              >
                <div
                  class="col-md-2 text-md-right font-weight-bold"
                >
                  <span
                    class="ngb-tooltip-class"
                    ngbtooltip="Message body of the notification."
                  >
                    Message content
                  </span>
                </div>
                <div
                  class="col-md-10 text-md-left"
                >
                  <tm-rich-text-editor
                    id="notification-message"
                  >
                    <div
                      inviewport=""
                    >
                    </div>
                  </tm-rich-text-editor>
                </div>
              </div>
              <br />
            </div>
          </div>
          <div
            class="card border-primary margin-top-20px"
          >
            <div
              class="card-body"
            >
              <div
                class="row text-center"
              >
                <div
                  class="col-md-2 text-md-right font-weight-bold"
                >
                  <span
                    class="ngb-tooltip-class"
                    ngbtooltip="This timezone is deduced based on your system/browser's settings."
                  >
                    Timezone
                  </span>
                </div>
                <div
                  class="col-md-4 text-md-left"
                  id="submission-status"
                >
                   Africa/Abidjan 
                </div>
              </div>
              <div
                class="row text-center margin-top-20px"
              >
                <div
                  class="col-md-6"
                >
                  <div
                    class="col-12 text-md-left"
                  >
                    <div>
                      <label
                        class="ngb-tooltip-class label-control font-weight-bold"
                        ngbtooltip="Users will be able to see the notification."
                      >
                         Notification start time 
                      </label>
                    </div>
                  </div>
                  <div
                    class="row text-center align-items-center"
                  >
                    <div
                      class="col-md-7 col-xs-center"
                      id="notification-start-date"
                    >
                      <tm-datepicker>
                        <div
                          class="input-group"
                        >
                          <input
                            class="form-control ng-untouched ng-pristine ng-valid"
                            ngbdatepicker=""
                            readonly=""
                            type="text"
                          />
                          <div
                            class="input-group-append"
                          >
                            <button
                              class="btn btn-light"
                              type="button"
                            >
                              <i
                                class="fas fa-calendar-alt"
                              />
                            </button>
                          </div>
                        </div>
                      </tm-datepicker>
                    </div>
                    <div
                      class="col-md-5"
                    >
                      <tm-timepicker
                        id="notification-start-time"
                      >
                        <select
                          class="form-control ng-untouched ng-pristine ng-valid"
                        >
                          <option
                            value="1: Object"
                          >
                             01:00H 
                          </option>
                          <option
                            value="2: Object"
                          >
                             02:00H 
                          </option>
                          <option
                            value="3: Object"
                          >
                             03:00H 
                          </option>
                          <option
                            value="4: Object"
                          >
                             04:00H 
                          </option>
                          <option
                            value="5: Object"
                          >
                             05:00H 
                          </option>
                          <option
                            value="6: Object"
                          >
                             06:00H 
                          </option>
                          <option
                            value="7: Object"
                          >
                             07:00H 
                          </option>
                          <option
                            value="8: Object"
                          >
                             08:00H 
                          </option>
                          <option
                            value="9: Object"
                          >
                             09:00H 
                          </option>
                          <option
                            value="10: Object"
                          >
                             10:00H 
                          </option>
                          <option
                            value="11: Object"
                          >
                             11:00H 
                          </option>
                          <option
                            value="12: Object"
                          >
                             12:00H 
                          </option>
                          <option
                            value="13: Object"
                          >
                             13:00H 
                          </option>
                          <option
                            value="14: Object"
                          >
                             14:00H 
                          </option>
                          <option
                            value="15: Object"
                          >
                             15:00H 
                          </option>
                          <option
                            value="16: Object"
                          >
                             16:00H 
                          </option>
                          <option
                            value="17: Object"
                          >
                             17:00H 
                          </option>
                          <option
                            value="18: Object"
                          >
                             18:00H 
                          </option>
                          <option
                            value="19: Object"
                          >
                             19:00H 
                          </option>
                          <option
                            value="20: Object"
                          >
                             20:00H 
                          </option>
                          <option
                            value="21: Object"
                          >
                             21:00H 
                          </option>
                          <option
                            value="22: Object"
                          >
                             22:00H 
                          </option>
                          <option
                            value="23: Object"
                          >
                             23:00H 
                          </option>
                          <option
                            value="0: Object"
                          >
                            23:59H
                          </option>
                        </select>
                      </tm-timepicker>
                    </div>
                  </div>
                </div>
                <div
                  class="col-md-6 border-left-gray"
                >
                  <div
                    class="row text-center"
                  >
                    <div
                      class="col-12 text-md-left ml-md-2"
                    >
                      <label
                        class="ngb-tooltip-class label-control font-weight-bold"
                        ngbtooltip="Notification will be hidden from users."
                      >
                         Notification end time 
                      </label>
                    </div>
                  </div>
                  <div
                    class="row align-items-center"
                  >
                    <div
                      class="col-md-7 col-xs-center"
                      id="notification-end-date"
                    >
                      <tm-datepicker>
                        <div
                          class="input-group"
                        >
                          <input
                            class="form-control ng-untouched ng-pristine ng-valid"
                            ngbdatepicker=""
                            readonly=""
                            type="text"
                          />
                          <div
                            class="input-group-append"
                          >
                            <button
                              class="btn btn-light"
                              type="button"
                            >
                              <i
                                class="fas fa-calendar-alt"
                              />
                            </button>
                          </div>
                        </div>
                      </tm-datepicker>
                    </div>
                    <div
                      class="col-md-5"
                    >
                      <tm-timepicker
                        id="notification-end-time"
                      >
                        <select
                          class="form-control ng-untouched ng-pristine ng-valid"
                        >
                          <option
                            value="1: Object"
                          >
                             01:00H 
                          </option>
                          <option
                            value="2: Object"
                          >
                             02:00H 
                          </option>
                          <option
                            value="3: Object"
                          >
                             03:00H 
                          </option>
                          <option
                            value="4: Object"
                          >
                             04:00H 
                          </option>
                          <option
                            value="5: Object"
                          >
                             05:00H 
                          </option>
                          <option
                            value="6: Object"
                          >
                             06:00H 
                          </option>
                          <option
                            value="7: Object"
                          >
                             07:00H 
                          </option>
                          <option
                            value="8: Object"
                          >
                             08:00H 
                          </option>
                          <option
                            value="9: Object"
                          >
                             09:00H 
                          </option>
                          <option
                            value="10: Object"
                          >
                             10:00H 
                          </option>
                          <option
                            value="11: Object"
                          >
                             11:00H 
                          </option>
                          <option
                            value="12: Object"
                          >
                             12:00H 
                          </option>
                          <option
                            value="13: Object"
                          >
                             13:00H 
                          </option>
                          <option
                            value="14: Object"
                          >
                             14:00H 
                          </option>
                          <option
                            value="15: Object"
                          >
                             15:00H 
                          </option>
                          <option
                            value="16: Object"
                          >
                             16:00H 
                          </option>
                          <option
                            value="17: Object"
                          >
                             17:00H 
                          </option>
                          <option
                            value="18: Object"
                          >
                             18:00H 
                          </option>
                          <option
                            value="19: Object"
                          >
                             19:00H 
                          </option>
                          <option
                            value="20: Object"
                          >
                             20:00H 
                          </option>
                          <option
                            value="21: Object"
                          >
                             21:00H 
                          </option>
                          <option
                            value="22: Object"
                          >
                             22:00H 
                          </option>
                          <option
                            value="23: Object"
                          >
                             23:00H 
                          </option>
                          <option
                            value="0: Object"
                          >
                            23:59H
                          </option>
                        </select>
                      </tm-timepicker>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="row margin-top-20px"
          >
            <div
              class="col-12 text-center"
            >
              <span
                class="loader"
              />
              <div>
                <button
                  class="btn btn-primary"
                  id="btn-create-notification"
                  type="button"
                >
                  Create Notification 
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </tm-notification-edit-form>
  </div><tm-loading-retry>
    <div
      class="margin-top-30px"
    >
      <div>
        <div
          class="alert alert-warning margin-top-30px"
          role="alert"
        >
           You have not created any notifications yet. Use the form above to create a notification. 
        </div>
      </div>
    </div>
    <br />
  </tm-loading-retry>
</tm-admin-notifications-page>
`;

exports[`AdminNotificationsPageComponent should snap when notifications are loading 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  currentNotificationEditFormMode="0"
  guessTimezone={[Function String]}
  hasNotificationLoadingFailed="false"
  isNotificationEditFormExpanded="false"
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in Africa/Abidjan time.
    </p>
  </div><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <span>
        <i
          class="fas fa-plus"
        />
         Add New Notification
      </span>
    </button>
  </div><tm-loading-retry>
    <div
      class="margin-top-30px"
    >
      <tm-loading-spinner>
        <div
          class="loading-container"
        >
          <div
            class="row spinner-border text-primary"
            role="status"
          />
          <div
            class="row text-secondary"
          >
            Loading...
          </div>
        </div>
      </tm-loading-spinner>
    </div>
    <br />
  </tm-loading-retry>
</tm-admin-notifications-page>
`;

exports[`AdminNotificationsPageComponent should snap when notifications failed to load 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  currentNotificationEditFormMode="0"
  guessTimezone={[Function String]}
  hasNotificationLoadingFailed={[Function Boolean]}
  isNotificationEditFormExpanded="false"
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in Africa/Abidjan time.
    </p>
  </div><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <span>
        <i
          class="fas fa-plus"
        />
         Add New Notification
      </span>
    </button>
  </div><tm-loading-retry>
    <div
      class="text-center"
    >
      <p
        class="text-muted m-1"
      >
        Failed to load data
      </p>
      <button
        class="btn btn-primary"
      >
        <i
          aria-hidden="true"
          class="fas fa-redo"
        />
        Â Retry
      </button>
    </div>
  </tm-loading-retry>
</tm-admin-notifications-page>
`;

exports[`AdminNotificationsPageComponent should snap with default view 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  currentNotificationEditFormMode="0"
  guessTimezone={[Function String]}
  hasNotificationLoadingFailed="false"
  isNotificationEditFormExpanded="false"
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in Africa/Abidjan time.
    </p>
  </div><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <span>
        <i
          class="fas fa-plus"
        />
         Add New Notification
      </span>
    </button>
  </div><tm-loading-retry>
    <div
      class="margin-top-30px"
    >
      <tm-loading-spinner>
        <div
          class="loading-container"
        >
          <div
            class="row spinner-border text-primary"
            role="status"
          />
          <div
            class="row text-secondary"
          >
            Loading...
          </div>
        </div>
      </tm-loading-spinner>
    </div>
    <br />
  </tm-loading-retry>
</tm-admin-notifications-page>
`;
